File: 2019/python/day01.py
import math
from functools import reduce


def calculate_fuel(module):
    fuel = math.floor(int(module) / 3) - 2

    return fuel + calculate_fuel(fuel) if fuel > 0 else 0


with open('inputs/day01.txt') as input:
    modules = input.readlines()
    total_fuel = reduce(lambda sum, module: sum +
                        calculate_fuel(module), modules, 0)
    print(total_fuel)


# Part one answer: 3296560

# Part two answer: 4941976

File: 2019/javascript/day01.js
const readInput = require('./read-input')

const input = readInput('day01.txt')
const modules = input.split('\n')

const calculateFuel = (module, recurse) => {
  const fuel = Math.floor(Number(module) / 3) - 2

  if (!recurse) return fuel

  return fuel > 0 ? fuel + calculateFuel(fuel) : 0
}

const partOne = modules.reduce((totalFuel, module) => totalFuel + calculateFuel(module), 0)
const partTwo = modules.reduce((totalFuel, module) => totalFuel + calculateFuel(module, true), 0)

console.log(partOne) // 3296560
console.log(partTwo) // 4941976
